Delay writing a terminal phase until the pod is terminated

smarterclayton opened the pull request:
Other components must know when the Kubelet has released critical resources for terminal pods. Do not set the phase in the apiserver to terminal until all containers are stopped and cannot restart.  Because the pod worker was inferring the transition to terminal state due to the propagation of the phase to the API and then back in from the config loop (which means that if the apiserver is down, failed and succeeded pods aren't torn down on the node), we need to instead have syncPod signal that the normal lifecycle of the pod has led to termination.

After those two changes, one of the init container tests was failing because TerminatePod was defaulting regular container status to the unknown error (137) even though we know for sure in this case no regular containers were ever started.  Correct TerminatePod defaulting to only include plausibly started containers (according to the pod lifecycle).  It's possible that the race that led to pods succeeding even though we never successfully ran the necessary containers is now fixed, but a future change will have to experiment with that.

Fixes #106884, #108733

#### What type of PR is this?

/kind bug
/kind regression

```release-note
Fixed a 1.22 regression that could incorrectly reject pods with OutOfCpu errors if they were rapidly scheduled after other pods were reported as complete in the API. The Kubelet now waits to report the phase of a pod as terminal in the API until all running containers are guaranteed to have stopped and no new containers can be started.  Short-lived pods may take slightly longer (~1s) to report Succeeded or Failed after this change.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

bobbypage committed:
test: Add E2E for job completions with cpu reservation

Create an E2E test that creates a job that spawns a pod that should
succeed. The job reserves a fixed amount of CPU and has a large number
of completions and parallelism. Use to repro github.com/kubernetes/kubernetes/issues/106884

Signed-off-by: David Porter <david@porter.me>

smarterclayton committed:
kubelet: Delay writing a terminal phase until the pod is terminated

Other components must know when the Kubelet has released critical
resources for terminal pods. Do not set the phase in the apiserver
to terminal until all containers are stopped and cannot restart.

As a consequence of this change, the Kubelet must explicitly transition
a terminal pod to the terminating state in the pod worker which is
handled by returning a new isTerminal boolean from syncPod.

Finally, if a pod with init containers hasn't been initialized yet,
don't default container statuses or not yet attempted init containers
to the unknown failure state.

aojea commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -654,6 +666,10 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
 }
 
+func isPhaseTerminal(phase v1.PodPhase) bool {
+	return phase == v1.PodFailed || phsae == v1.PodSucceeded
```
```suggestion
	return phase == v1.PodFailed || phase == v1.PodSucceeded
```

bobbypage commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -619,6 +621,16 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
+	// complete. The Kubelet should never transition a pod to terminal status that could
+	// have running containers and thus actively be leveraging exclusive resources.
+	if !isPhaseTerminal(pod.Status.Phase) && isPhaseTerminal(status.status.Phase) {
+		if m.podDeletionSafety.PodCouldHaveRunningContainers(pod) {
```
Just to understand why are we only using `PodCouldHaveRunningContainers` which only checks `CouldHaveRunningContainers` instead of [`PodResourcesAreReclaimed`](https://github.com/kubernetes/kubernetes/blob/043611723136d30ec76782f96c5bb03fe7a47ca4/pkg/kubelet/kubelet_pods.go#L920)  which is a stronger check that also validates that all containers are exited, no pod volumes exist, cgroup is deleted etc?

smarterclayton replied:
PodResourcesAreReclaimed is “fully terminated”, ie syncTerminatedPod has completed.  The issue in this case is terminating containers, which corresponds with exiting syncTerminatingPods.  As mentioned in slack, if you start a new pod with volumes while volumes used by a previous pod are cleaning up, we gracefully wait already.  So there’s no impact to an end user by accepting that pod early.  We actually need that behavior for certain types of zero disruption upgrades of things that are per node.

The reclaimed check is to cover deletion of a pod, where we need to ensure another node can safely and exclusively access some volume types - unmount ensures no more data can be written to that disk (well, it gives the volume plugin the chance to provide that semantic).  If we didn’t wait, we wouldn’t guarantee the pod safety rule that a volume the scheduler believes is uniquely assigned to a specific node can’t accept writes from another node.

smarterclayton replied:
Added an extended comment to this effect in the updated PR.  Let me know if that's sufficiently clear (I plan on trying to boil this into a KEP that clarifies the current state and proposes the improvements that could be added, like additional conditions).

aojea commented:
:eyes:  lot of these entries https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1498078380036198400/artifacts/e2e-8e4edba04b-a7d53-minion-group-72c3/kubelet.log

```
eb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]: I0228 00:44:32.631401   10034 status_manager.go:715] "Pod status is inconsistent with cached status for pod, a reconciliation should be triggered" pod="secrets-3867/pod-secrets-0464bf08-aa94-4d04-bd2b-0960c97a67a3" statusDiff=<
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:           &v1.PodStatus{
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:         -         Phase:      "Pending",
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:         +         Phase:      "Succeeded",
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:                   Conditions: {{Type: "Initialized", Status: "True", LastTransitionTime: {Time: s"2022-02-28 00:44:13 +0000 UTC"}, Reason: "PodCompleted", ...}, {Type: "Ready", Status: "False", LastTransitionTime: {Time: s"2022-02-28 00:44:13 +0000 UTC"}, Reason: "PodCompleted", ...}, {Type: "ContainersReady", Status: "False", LastTransitionTime: {Time: s"2022-02-28 00:44:13 +0000 UTC"}, Reason: "PodCompleted", ...}, {Type: "PodScheduled", Status: "True", LastTransitionTime: {Time: s"2022-02-28 00:44:13 +0000 UTC"}}},
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:                   Message:    "",
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:                   ... // 4 identical fields
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:                   PodIPs:                     {{IP: "10.64.3.171"}},
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:                   StartTime:                  s"2022-02-28 00:44:13 +0000 UTC",
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:         -         InitContainerStatuses:      nil,
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:         +         InitContainerStatuses:      []v1.ContainerStatus{},
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:                   ContainerStatuses:          {{Name: "secret-env-test", State: {Terminated: &{Reason: "Completed", StartedAt: {Time: s"2022-02-28 00:44:14 +0000 UTC"}, FinishedAt: {Time: s"2022-02-28 00:44:14 +0000 UTC"}, ContainerID: "containerd://31c4d1c2d263d5ec36521a892e0d848178c78e1aab9409b0c63"...}}, Image: "k8s.gcr.io/e2e-test-images/busybox:1.29-2", ImageID: "k8s.gcr.io/e2e-test-images/busybox@sha256:c318242786b139d18676b1"..., ...}},
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:                   QOSClass:                   "BestEffort",
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:         -         EphemeralContainerStatuses: nil,
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:         +         EphemeralContainerStatuses: []v1.ContainerStatus{},
Feb 28 00:44:32.631423 e2e-8e4edba04b-a7d53-minion-group-72c3 kubelet[10034]:           }
```

249043822 commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -771,25 +774,47 @@ func normalizeStatus(pod *v1.Pod, status *v1.PodStatus) *v1.PodStatus {
+	// the Kubelet exclusively owns must be released prior to a pod being reported terminal,
+	// while resources that have participanting components above the API use the pod's
+	// transition to a terminal phase (or full deletion) to release those resources.
+	if !isPhaseTerminal(oldPodStatus.Phase) && isPhaseTerminal(newPodStatus.Phase) {
```
This change means that we would sync `Running `phase to apiserver while the pod has the phase `Succeed` but it's still Terminating on kubelet, then the podstatus in statusmanager is inconsistent with that in apiserver for a brief time.
I have another idea that can we introduce a intermediate phase like ` "Terminating"` which would not affect the existing code logic on scheduler and controller-manager, only make kubelet identify this phase, so while the pod should have the phase `Succeed` but it's still Terminating on kubelet, we first set the phase `Terminating`, after the containers are all shutted down, then the pod phase becomes `Succeed `or `Failed`, can this apporach solve the isssue?

smarterclayton commented:
Directly described the transition from a syncPod where the pod is known terminal into syncTerminatingPod.  That should address the failure to transition described in the previous failures (and is a necessary consequence, since the Kubelet "knows" that we transitioned to terminating, but was instead relying on a status update to propagate it)

bobbypage commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -915,6 +915,12 @@ func countRunningContainerStatus(status v1.PodStatus) int {
+// PodCouldHaveRunningContainers returns true if the pod with the given UID could still have running
+// containers. This returns false if the pod has not yet been started or the pod is unknown.
+func (kl *Kubelet) PodCouldHaveRunningContainers(pod *v1.Pod) bool {
+	return kl.podWorkers.CouldHaveRunningContainers(pod.UID)
```
Do we also need to check the status to ensure no containers are running? i.e. the check here - https://github.com/kubernetes/kubernetes/blob/f4df7d0eb27b51a0f47a49616b8b32e0bf9c18a7/pkg/kubelet/kubelet_pods.go#L926-L931

I'm not sure if this check is redundant or not, but at least in `PodResourcesAreReclaimed` it checks both `podWorkers.CouldHaveRunningContainers` as well as the running containers from the status...

smarterclayton replied:
The check in PodResourcesReclaimed isn't checking for running containers - it's checking to see that we have received all the container statuses from the status loop before we delete the pod (we're trying to make sure we propagate all the expected statuses).  TerminatePod called within syncTerminatedPod guarantees that those statuses are flushed - the wait guarantees we write them to the status before we finally delete the pod (which is the last possible time to write the container status).

PodCouldHaveRunning containers is weaker - we don't need to be able to write all the container status before we write the phase, as long as it is written later.  So we need to wait until those statuses propagate.

249043822 commented on pkg/kubelet/pod_workers.go:
```diff
@@ -1115,9 +1151,11 @@ func (p *podWorkers) completeUnstartedTerminated(pod *v1.Pod) {
 	// Requeue the last update if the last sync returned error.
 	switch {
+	case phaseTransition:
+		p.workQueue.Enqueue(pod.UID, 0)
```
want the next sync for phase transition as soon as possible, because the syncErr == nil, so it would not reach default case, right?

249043822 replied:
I have a little question that if  `Enqueue `delay set 0, is there a extreme race that  `UpdatePod` for "phaseTransition" precedes `p.completeWorkQueueNext(pod.UID)` for "TerminatingPodWork"?  If this case, phaseTransition may lose efficacy, Of course, this possibility should be minimal.

smarterclayton replied:
The original state of the pod worker did not maintain a state machine (and hence we had all sorts of races and bugs).  We added the state machine, but preserved some of the "interesting" additions - specifically status.GetNewerThan, which tries not to call syncPod on the same container status more than once.  However, phase transitions can occur when no new status updates could be sent (ever), so we don't need to wait any longer than it is required for the periodic pod status refresh to trigger the cache.  There's a long todo above GetNewerThan, but roughly in the future we want sync*Pod to be able to trigger a status refresh and to instantly make a transition (right now we always wait 1s because of various reasons).  So this is starting to put in place that mechanism, because we have no need to wait at all.

bobbypage replied:
Makes sense, thanks for the clarification.

smarterclayton commented:
Working through the status behavior in the latest commit, I *think* after the changes to status here, we may no longer need TerminatePod to default the status (the bug that originally led to status being reported was due to this race between pod status and pod phase), but that's a larger change.  Also, #107845 probably has to merge first, because failures in CRI could hold it.

smarterclayton commented on pkg/kubelet/status/status_manager_test.go:
```diff
@@ -648,11 +649,14 @@ func TestTerminatePodWaiting(t *testing.T) {
 	newStatus := expectPodStatus(t, syncer, testPod)
-	for i := range newStatus.ContainerStatuses {
-		assert.False(t, newStatus.ContainerStatuses[i].State.Terminated == nil, "expected containers to be terminated")
+	for _, container := range newStatus.ContainerStatuses {
```
@ehashman this change is deliberate on top of your https://github.com/kubernetes/kubernetes/pull/105462 change, and ensures that there is always at least one terminated container with non-zero exit in case any info is missing (if there are no init containers, all containers get the behavior, if we can't prove we initialized fully, then the last init container is failed).  See changes to TerminatePod()

I need to triple check that a pod with all waiting containers and which has succeeded initialization is listed as failed, but that's the lower test in this file.

smarterclayton commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -335,19 +337,26 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
-				ExitCode: 137,
-			},
+
+	// once a pod has initialized, any missing status is treated as a failure
```
Changes here should preserve the invariant that at least one container is failed in the pod (Init or regular) or we mark all waiting containers as terminated with 137.  Reason being that hasPodInitialized is true if the last init container is successful or there are no init containers, OR if there is a failing init container (which for restart never pods is enough to mark as failure).

The e2e in #107829 should be sufficient to test this didn't regress, but I may resurrect a test of a similar behavior on init containers.

bobbypage replied:
> Changes here should preserve the invariant that at least one container is failed in the pod (Init or regular) or we mark all waiting containers as terminated with 137. 

The current code checks

```
if state := container.State; state.Terminated != nil && state.Terminated.ExitCode == 0 {
			return true
}
```

so if exitCode != 0, the `hasPodInitialized` will return false. This seems to be contradict the statement:

> hasPodInitialized is true if the last init container is successful or there are no init containers, **OR if there is a failing init container**

So in the current case, AFAIU, if there is failing init container, `hasPodInitialized` will return false, and we will not enter the body of the loop and set rest of container status to exit code 137? is that the expected behavior?

smarterclayton commented:
/retest

rphillips commented:
Testing this PR and [PR#107845](https://github.com/kubernetes/kubernetes/pull/107845) in combination within OpenShift CI. [PR](https://github.com/openshift/kubernetes/pull/1199)

rphillips commented:
@smarterclayton wdyt o merging https://github.com/kubernetes/kubernetes/pull/107845 into this PR to ease in backporting.

smarterclayton commented:
#107845 is wrong, but isn't related to this PR directly (or at least, I don't have any evidence it's in play here).

smarterclayton commented:
Units are not failing for me locally, I don't know why:

```
=== RUN   TestTerminatePod_DefaultUnknownStatus/#00
    status_manager_test.go:979: unexpected status: 
    --- FAIL: TestTerminatePod_DefaultUnknownStatus/#00 (0.00s)
```

looking into that too

smarterclayton commented:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/job.go:371
Mar  1 19:01:35.564: expected 100 successful job pods, but got  99
Expected
    <int32>: 99
to equal
    <int32>: 100
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/job.go:414
```

in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1498722885781426176/ is likely related to these changes, looking.

smarterclayton commented:
/retest

bobbypage commented:
In the latest CI run,

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1498722885781426176/

the failure of 

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/job.go:371
Mar  1 19:01:35.564: expected 100 successful job pods, but got  99
Expected
    <int32>: 99
to equal
    <int32>: 100
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/job.go:414
```

appears to be due to missing pod `all-succeed-46d9p`

The pod is created (and container is started) but status of being in succeeded is missing from the dump of pods at the end of the [build-log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1498722885781426176/build-log.txt)
```
I0301 19:01:43.214] Mar  1 19:01:35.658: INFO: At 2022-03-01 18:59:37 +0000 UTC - event for all-succeed-46d9p: {kubelet e2e-b710a6b5d6-a7d53-minion-group-ps83} Started: Started container c
```

rphillips commented:
Looks like the controller-manager's pod-garbage-collector deleted `all-succeed-46d9p`:

[kube-apiserver.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1498722885781426176/artifacts/e2e-b710a6b5d6-a7d53-master/kube-apiserver.log):
```
│I0301 19:01:35.031527       9 store.go:298] deletion of /registry/pods/job-4605/all-succeed-46d9p failed because of a conflict, going to retry                                                                                                                                                                                                                                                                                             │
│I0301 19:01:35.098831       9 httplog.go:131] "HTTP" verb="DELETE" URI="/api/v1/namespaces/job-4605/pods/all-succeed-46d9p" latency="191.163008ms" userAgent="kube-controller-manager/v1.24.0 (linux/amd64) kubernetes/a8c255b/system:serviceaccount:kube-system:pod-garbage-collector" audit-ID="8ad61fe1-6f91-4f8a-806c-39d6f7ab4dd2" srcIP="127.0.0.1:33910" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_fd="job-46│
│I0301 19:01:35.113016       9 graph_populator.go:180] deletePod job-4605/all-succeed-46d9p for node e2e-b710a6b5d6-a7d53-minion-group-ps83   
```

bobbypage commented:
> Looks like the controller-manager's pod-garbage-collector deleted `all-succeed-46d9p`:
> 
> [kube-apiserver.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1498722885781426176/artifacts/e2e-b710a6b5d6-a7d53-master/kube-apiserver.log):
> 
> ```
> │I0301 19:01:35.031527       9 store.go:298] deletion of /registry/pods/job-4605/all-succeed-46d9p failed because of a conflict, going to retry                                                                                                                                                                                                                                                                                             │
> │I0301 19:01:35.098831       9 httplog.go:131] "HTTP" verb="DELETE" URI="/api/v1/namespaces/job-4605/pods/all-succeed-46d9p" latency="191.163008ms" userAgent="kube-controller-manager/v1.24.0 (linux/amd64) kubernetes/a8c255b/system:serviceaccount:kube-system:pod-garbage-collector" audit-ID="8ad61fe1-6f91-4f8a-806c-39d6f7ab4dd2" srcIP="127.0.0.1:33910" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_fd="job-46│
> │I0301 19:01:35.113016       9 graph_populator.go:180] deletePod job-4605/all-succeed-46d9p for node e2e-b710a6b5d6-a7d53-minion-group-ps83   
> ```

ah, that would explain it. Looks like kube-controller-managed is configured to have 100 be the threshold (I'm not sure why it's so low when the default is 12500) - https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/

```
artifacts/e2e-b710a6b5d6-a7d53-master/kube-controller-manager.log
143:I0301 18:39:42.732386       8 flags.go:64] FLAG: --terminated-pod-gc-threshold="100"
```

rphillips commented:
@bobbypage [pod_gc.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/pod_gc.go#L34) requires a setting of 100 for the e2e.

rphillips commented:
@smarterclayton updating this PR's [count](https://github.com/kubernetes/kubernetes/pull/108366/files#diff-909ebfa7fcb2a94030da3371d35863a2d723fc93356334e35f68816f5e313379R51-R52) to be 75 or so should clear up the issue.

smarterclayton commented:
Looking at 

```
Kubernetes e2e suite: [sig-node] Pods should delete a collection of pods [Conformance] expand_less | 5m10s
-- | --
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:633 Mar  1 20:00:01.734: found a pod(s) Unexpected error:     <*errors.errorString \| 0xc00027c240>: {         s: "timed out waiting for the condition",     }     timed out waiting for the condition occurred /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/node/pods.go:886
```

which may be unrelated - I just debugged something similar in an unrelated PR (that doesn't touch pod status) https://github.com/kubernetes/kubernetes/pull/102025#issuecomment-1055842724

smarterclayton commented:
@rphillips do the jobs run in parallel on the same node?  if so, that may be an issue even at 75 (conversely, the gc being set to a 100 for pod gc testing may be too low).

rphillips commented:
Good point. The Job E2E's are using 100, but in this case we don't want the pod to be GCed (those tests are not checking for this use case).

rphillips commented:
This might fix it... I can't preproduce the error locally.

```diff
diff --git a/pkg/kubelet/status/status_manager_test.go b/pkg/kubelet/status/status_manager_test.go
index c9105ea60e6..51b13258958 100644
--- a/pkg/kubelet/status/status_manager_test.go
+++ b/pkg/kubelet/status/status_manager_test.go
@@ -697,7 +697,8 @@ func TestTerminatePod_DefaultUnknownStatus(t *testing.T) {
 				Name: fmt.Sprintf("%d", i),
 			})
 		}
-		pod.Status.StartTime = &metav1.Time{Time: time.Unix(1, 0)}
+		now := metav1.Time{Time: time.Unix(1, 0)}.Rfc3339Copy()
+		pod.Status.StartTime = &now
 		for _, fn := range fns {
 			fn(pod)
 		}
```

smarterclayton commented:
Yeah, a bug in cmp.Diff was hiding it (UTC and local are the same), I have a slightly different tweak (but thanks).

smarterclayton commented:
Failure in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-conformance-kind-ga-only-parallel/1498776684252368896

```
Kubernetes  e2e suite: [sig-node] Probing container should have monotonically  increasing restart count [NodeConformance] [Conformance] expand_less | 6m24s
-- | --
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:633 Mar  1 22:16:30.908: pod container-probe-4257/liveness-b8e4b0bb-d9ed-4102-a738-c994ca9f9ace - expected number of restarts: 5, found restarts: 4 /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/node/container_probe.go:195
```

Looks like is happening outside of this PR https://storage.googleapis.com/k8s-triage/index.html?test=should%20have%20monotonically%20increasing%20restart%20count.  I don't see a bug open for it, but it's roughly that exponential restart backoff exceeds the 5m test timeout if the test takes too long to grab the first pod status.   I don't think it's related to this PR at all.

smarterclayton commented:
In https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-conformance-kind-ipv6-parallel/1498776675930869760

```
Kubernetes  e2e suite: [sig-node] Variable Expansion should fail substituting  values in a volume subpath with backticks [Slow] [Conformance] expand_less | 6m9s
-- | --
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:633 Mar  1 22:16:19.767: while waiting for the pod container to fail Unexpected error:     <*errors.errorString \| 0xc00020e250>: {         s: "timed out waiting for the condition",     }     timed out waiting for the condition occurred /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/node/expansion.go:379
```

is possibly related to the change.  The test creates a pod that should be "waiting" with an error message.

The first syncPod exits with the error

```
Mar 01 22:14:53 kind-worker2 kubelet[259]: I0301 22:14:51.547558     259 kubelet.go:1502] "syncPod exit" pod="var-expansion-2335/var-expansion-f497ce22-9923-4251-9dec-a13fdd8c839f" podUID=64fc0dbc-e1b0-4755-96f6-cf0a1ce57f58 isTerminal=false
Mar 01 22:14:53 kind-worker2 kubelet[259]: E0301 22:14:51.547576     259 pod_workers.go:951] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"dapi-container\" with CreateContainerConfigError: \"unable to provision SubPath `..`: must not contain '..'\"" pod="var-expansion-2335/var-expansion-f497ce22-9923-4251-9dec-
```

But then the next syncPod doesn't enter for 3m:

```
Mar 01 22:17:45 kind-worker2 kubelet[259]: I0301 22:17:42.811494     259 kubelet.go:1500] "syncPod enter" pod="var-expansion-2335/var-expansion-f497ce22-9923-4251-9dec-a13fdd8c839f" podUID=64fc0dbc-e1b0-4755-96f6-cf0a1ce57f58
```

and then the status update is ignored (?!?)

```
Mar 01 22:17:45 kind-worker2 kubelet[259]: I0301 22:17:44.562953     259 status_manager.go:535] "Ignoring same status for pod" pod="var-expansion-2335/var-expansion-f497ce22-9923-4251-9dec-a13fdd8c839f" status={Phase:Pending Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-
```

Something is very, very, very wrong.

smarterclayton commented:
Latest failure in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-conformance-kind-ipv6-parallel/1499135865555783680 is not this PR, https://storage.googleapis.com/k8s-triage/index.html?test=%20A%20set%20of%20valid%20responses%20are%20returned%20for%20both%20pod%20and%20service%20Proxy#3ec994cd7076ec1bbe82

smarterclayton commented:
Failure in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1499135865639669760/

```
Kubernetes  e2e suite: [sig-storage] CSI Volumes [Driver: csi-hostpath]  [Testpattern: Generic Ephemeral-volume (default fs)] volumeLimits should  verify that all csinodes have volume limits expand_less | 1m18s
-- | --
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumelimits.go:247 Mar  2 22:13:00.176: Expected volume limits to be set, error: could not get CSINode limit for driver csi-hostpath: timed out waiting for the condition /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
```

Looks like a flake outside this PR

https://storage.googleapis.com/k8s-triage/index.html?text=error%3A%20could%20not%20get%20CSINode%20limit%20for%20driver%20csi-hostpath#6f6d6d5a53a81db58335

249043822 commented:
I merged this pr to our branch, the `outOfCPU `could not repro again,  looking logs...
```
root@ubuntu:/home/ubuntu# kubectl get po
NAME                    READY   STATUS              RESTARTS          AGE
busybox--1-2h7k5        0/1     Completed           0                 70s
busybox--1-2hjj8        0/1     Completed           0                 7m1s
busybox--1-2xr8q        0/1     Completed           0                 4m5s
busybox--1-2zkv5        0/1     Pending             0                 2m26s
busybox--1-422kz        0/1     Pending             0                 7m1s
busybox--1-4bxrn        0/1     Completed           0                 6m44s
busybox--1-4fzpp        0/1     Completed           0                 5m53s
busybox--1-4jkxg        0/1     Completed           0                 70s
busybox--1-56hvx        0/1     Pending             0                 7m
busybox--1-58gf9        0/1     Pending             0                 7m1s
busybox--1-5gz48        0/1     Completed           0                 4m25s
busybox--1-5ll4c        0/1     Completed           0                 3m46s
busybox--1-5qnnf        0/1     Completed           0                 2m11s
busybox--1-642rr        0/1     Completed           0                 109s
busybox--1-6d7dg        0/1     Completed           0                 6m28s
busybox--1-6dlmj        0/1     Completed           0                 7m1s
busybox--1-6lsmb        0/1     Pending             0                 7m1s
busybox--1-6s68x        0/1     Completed           0                 6m20s
busybox--1-6xrqd        0/1     Completed           0                 52s
busybox--1-6zq5c        0/1     Completed           0                 3m30s
busybox--1-79hcb        0/1     Pending             0                 7m1s
busybox--1-7jrzl        0/1     Completed           0                 64s
busybox--1-7p7dg        0/1     Completed           0                 3m8s
busybox--1-7rxbw        0/1     Completed           0                 4m3s
busybox--1-7t4wk        0/1     Pending             0                 7m
busybox--1-8884f        0/1     Completed           0                 3m52s
busybox--1-8cvp8        0/1     Pending             0                 14s
busybox--1-8nlt8        0/1     Completed           0                 3m30s
busybox--1-8sgd6        0/1     Completed           0                 24s
busybox--1-8snxh        0/1     Completed           0                 6m9s
busybox--1-8xk52        0/1     Completed           0                 7m1s
busybox--1-92m42        0/1     Completed           0                 4m53s
busybox--1-97c8x        0/1     Completed           0                 2m41s
busybox--1-9hgwq        0/1     Completed           0                 6m4s
busybox--1-9jvrw        0/1     Completed           0                 3m14s
busybox--1-9llzs        0/1     Completed           0                 4m29s
busybox--1-9lsz9        0/1     Completed           0                 7m1s
busybox--1-9tzdk        0/1     Completed           0                 6m40s
busybox--1-9v9xw        0/1     Completed           0                 4m43s
busybox--1-9zldx        0/1     Completed           0                 81s
busybox--1-b5972        0/1     Pending             0                 7m1s
busybox--1-b8p6t        0/1     Completed           0                 109s
busybox--1-b9mqp        0/1     Completed           0                 5m12s
busybox--1-b9n9q        0/1     Pending             0                 7m1s
busybox--1-bd2x6        0/1     Completed           0                 2m38s
busybox--1-bd78v        0/1     Completed           0                 6m44s
busybox--1-bg2wj        0/1     Completed           0                 3m14s
busybox--1-bgdkp        0/1     Completed           0                 4m49s
busybox--1-bpnk8        0/1     Completed           0                 42s
busybox--1-br8cz        0/1     Pending             0                 7m
busybox--1-brn6f        0/1     Completed           0                 2m54s
busybox--1-bxw78        0/1     Completed           0                 14s
busybox--1-c4ptl        0/1     Completed           0                 2m54s
busybox--1-c4vqb        0/1     Completed           0                 61s
busybox--1-c7qnk        0/1     Completed           0                 3m44s
```

smarterclayton commented:
/retest

smarterclayton commented:
/test all

smarterclayton commented:
/test pull-kubernetes-node-kubelet-serial-containerd

smarterclayton commented:
Any other test types people can think of?  Every flake so far is unrelated to this, and while this increases latency between pod actually finishes and pod being reported as finished, that's intentional (the previous value was negative latency between finished and correct finished).

Do we need more soak with reproducers of outofcpu beyond the e2e test and what @249043822 has done?

smarterclayton commented:
/test all

smarterclayton commented:
/retest

bobbypage commented:
> Any other test types people can think of? Every flake so far is unrelated to this, and while this increases latency between pod actually finishes and pod being reported as finished, that's intentional (the previous value was negative latency between finished and correct finished).

I definitely think this is right direction of the fix, we should delay writing terminal phase even if it increases latency since the previously kubelet reported terminal phase too early which wasn't accurate in the first place. 

Regarding tests, I think the e2e out of cpu error should capture the main fix to the issue since it was consistently failing without this fix. Do we have sufficient testing of container status changes, i.e. when waiting containers/init containers should be set to exit code 137 (i.e. your comment here - https://github.com/kubernetes/kubernetes/pull/108366/files#r817033842). Perhaps that's what we need to add...

bobbypage commented:
I did a manual test with below pod, and I think there is an issue with reporting Failed state for main containers when the init containers don't succeed:

Here's the pod:
```
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
spec:
  restartPolicy: Never
  containers:
  - name: main-container
    image: busybox:1.28
    command: ['/bin/false']
  initContainers:
  - name: init-1
    image: busybox:1.28
    command: ['/bin/true']
  - name: init-2
    image: busybox:1.28
    command: ['/bin/false']
```

After this change, this is status (main-container is stuck on `PodInitializing` which doesn't look to be right):

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:21:12Z"
    message: 'containers with incomplete status: [init-2]'
    reason: ContainersNotInitialized
    status: "False"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:21:12Z"
    message: 'containers with unready status: [main-container]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:21:12Z"
    message: 'containers with unready status: [main-container]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:21:12Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: busybox:1.28
    imageID: ""
    lastState: {}
    name: main-container
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        reason: PodInitializing
  hostIP: 192.168.10.2
  initContainerStatuses:
  - containerID: containerd://21460e6550583af6e7df7a7141ed0f5d3205af347893911fd07dccbde5078ae5
    image: docker.io/library/busybox:1.28
    imageID: docker.io/library/busybox@sha256:141c253bc4c3fd0a201d32dc1f493bcf3fff003b6df416dea4f41046e0f37d47
    lastState: {}
    name: init-1
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: containerd://21460e6550583af6e7df7a7141ed0f5d3205af347893911fd07dccbde5078ae5
        exitCode: 0
        finishedAt: "2022-03-07T23:21:15Z"
        reason: Completed
        startedAt: "2022-03-07T23:21:15Z"
  - containerID: containerd://8a4b9249d0ea9f52aed301c7cd4d561e988822d04f19fa75441faa4a077fa162
    image: docker.io/library/busybox:1.28
    imageID: docker.io/library/busybox@sha256:141c253bc4c3fd0a201d32dc1f493bcf3fff003b6df416dea4f41046e0f37d47
    lastState: {}
    name: init-2
    ready: false
    restartCount: 0
    state:
      terminated:
        containerID: containerd://8a4b9249d0ea9f52aed301c7cd4d561e988822d04f19fa75441faa4a077fa162
        exitCode: 1
        finishedAt: "2022-03-07T23:21:16Z"
        reason: Error
        startedAt: "2022-03-07T23:21:16Z"
  phase: Failed
  podIP: 10.244.1.2
  podIPs:
  - ip: 10.244.1.2
  qosClass: BestEffort
  startTime: "2022-03-07T23:21:12Z"
```

Creating same test on cluster on master without this change, this is the state:

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:29:37Z"
    message: 'containers with incomplete status: [init-2]'
    reason: ContainersNotInitialized
    status: "False"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:29:37Z"
    message: 'containers with unready status: [main-container]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:29:37Z"
    message: 'containers with unready status: [main-container]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2022-03-07T23:29:37Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: busybox:1.28
    imageID: ""
    lastState: {}
    name: main-container
    ready: false
    restartCount: 0
    started: false
    state:
      terminated:
        exitCode: 137
        finishedAt: null
        message: The container could not be located when the pod was terminated
        reason: ContainerStatusUnknown
        startedAt: null
  hostIP: 192.168.10.2
  initContainerStatuses:
  - containerID: containerd://24522d5251353280e5ad39a1a67ae57333b9b10918164a4223188fcb46b312c9
    image: docker.io/library/busybox:1.28
    imageID: docker.io/library/busybox@sha256:141c253bc4c3fd0a201d32dc1f493bcf3fff003b6df416dea4f41046e0f37d47
    lastState: {}
    name: init-1
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: containerd://24522d5251353280e5ad39a1a67ae57333b9b10918164a4223188fcb46b312c9
        exitCode: 0
        finishedAt: "2022-03-07T23:29:41Z"
        reason: Completed
        startedAt: "2022-03-07T23:29:41Z"
  - containerID: containerd://7a0882cfaa08cc9ad4de5a4f9ae71583dfe9ca510d10f7834205596240c411ea
    image: docker.io/library/busybox:1.28
    imageID: docker.io/library/busybox@sha256:141c253bc4c3fd0a201d32dc1f493bcf3fff003b6df416dea4f41046e0f37d47
    lastState: {}
    name: init-2
    ready: false
    restartCount: 0
    state:
      terminated:
        containerID: containerd://7a0882cfaa08cc9ad4de5a4f9ae71583dfe9ca510d10f7834205596240c411ea
        exitCode: 1
        finishedAt: "2022-03-07T23:29:41Z"
        reason: Error
        startedAt: "2022-03-07T23:29:41Z"
  phase: Failed
  podIP: 10.244.1.2
  podIPs:
  - ip: 10.244.1.2
  qosClass: BestEffort
  startTime: "2022-03-07T23:29:37Z"
```

(pod is marked with exit code 137).

bobbypage commented:
Additionally, as discussed over slack, there is a related issue with terminal pods - https://github.com/kubernetes/kubernetes/issues/108594 where a terminal pod status update can report Phase:Failed, Ready=true which shouldn't be a valid status update. I reproduced this with the current implementation, and the fix here does not currently address that issue.

bobbypage commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -619,8 +679,9 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
 
-	oldStatus := pod.Status.DeepCopy()
-	newPod, patchBytes, unchanged, err := statusutil.PatchPodStatus(m.kubeClient, pod.Namespace, pod.Name, pod.UID, *oldStatus, mergePodStatus(*oldStatus, status.status))
+	mergedStatus := mergePodStatus(pod.Status, status.status, m.podDeletionSafety.PodCouldHaveRunningContainers(pod))
```
I think we should change the log line on line 694 to 

```
klog.V(3).InfoS("Status for pod updated successfully", "pod", klog.KObj(pod), "statusVersion", status.version, "status", mergedStatus)
```

since we are incorrectly logging `status.status` as the status that was updated, when it actually is mergedStatus

smarterclayton replied:
Thank you, I think that was one of my confusion points in debugging.

bobbypage committed:
test: Verify that nodes do not transition to Failed while ready

Signed-off-by: David Porter <david@porter.me>

rphillips commented:
I tried this in [OpenShift CI](https://github.com/openshift/kubernetes/pull/1199). There appears to still be an issue with pod statuses still [[CI run]](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_kubernetes/1199/pull-ci-openshift-kubernetes-master-k8s-e2e-gcp/1501264553625784320).

I'll bump the logging so we can get more information.

smarterclayton commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -364,6 +381,49 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
+			return true
+		}
+	}
+	// if the last init container has not completed with a zero exit code
```
Agree the comment is confusing, because I can't even answer without looking at code :)

smarterclayton commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -335,19 +345,26 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
+
+	// all but the final suffix of init containers which have no evidence of a container start are
+	// marked as failed containers
+	for i := range initializedContainers(status.InitContainerStatuses) {
```
This is tricky.  The presence of an init container at `N+1` implies that any previous init container exited with code 0 at least one time.  Therefore by inductive reasoning, it would be wrong to transform `[<uninit>, exit=0] -> [exit=137, exit=0]`.  *Technically* `[<uninit>, exit=0] -> [exit=0, exit=0]` is actually correct, but I'm nervous about synthesizing the info there.

The code here is supposed to handle `[..., <uninit>] -> [..., exit=137]`

bobbypage replied:
ah, that makes sense. I think the larger point is the same as you mentioned in https://github.com/kubernetes/kubernetes/pull/108366#issuecomment-1063278740.

If the init container is not initialized, we don't want to set the exit code to 137. We should only set exit code to 137 if we have confirmed that the container has actually started. As a result with this change, now, e.g. if you have three init containers and first succeeds, second fails, the third will stay on "waiting" as opposed to having exit code 137 set.

smarterclayton commented:
Re: https://github.com/kubernetes/kubernetes/pull/108366#issuecomment-1061248318

The change to leave the container in waiting is intentional - the pod is failed (initialization didn't succeed), the container could *never* have been started (initialization didn't pass).  So that is a deliberate consequence of my change - we shouldn't say the regular container exited if it cannot be proven to have started (we have no evidence it started, which means at the time the pod failed the container was 'waiting').

We have had a lot of bugs because containers that never really terminated were listed as accidentally exitCode=0 (the default), which accidentally led to succeeded pods (that didn't clearly succeed).  I think of this change as "we can't set `terminated != nil` unless we actually observed a start, or we can't prove we didn't start it at some point and then lost the status".

bobbypage commented:
The test change in https://github.com/kubernetes/kubernetes/pull/108599 can be cherrypicked into this PR as an additional test case to test fix for https://github.com/kubernetes/kubernetes/issues/108594. In my local test of this node e2e, with this PR the test no longer fails, but it fails in master (as expected).

bobbypage commented:
> We have had a lot of bugs because containers that never really terminated were listed as accidentally exitCode=0 (the default), which accidentally led to succeeded pods (that didn't clearly succeed). I think of this change as "we can't set `terminated != nil` unless we actually observed a start, or we can't prove we didn't start it at some point and then lost the status".

I see, thanks for clarification, this makes sense. This does seem like a behavior change though and will be worth to document somewhere, but I agree it does seem like the proper behavior from kubelet to leave the container as waiting if initialization did not complete due to init containers failing to start.

bobbypage commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -771,25 +832,47 @@ func normalizeStatus(pod *v1.Pod, status *v1.PodStatus) *v1.PodStatus {
+	// transition to a terminal phase (or full deletion) to release those resources.
+	if !isPhaseTerminal(oldPodStatus.Phase) && isPhaseTerminal(newPodStatus.Phase) {
+		if couldHaveRunningContainers {
+			newPodStatus.Phase = oldPodStatus.Phase
```
I think we may have to should preserve the old message and reason as well:

```
newPodStatus.Message = oldPodStatus.Message
newPodStatus.Reason = oldPodStatus.Reason
```

I did some testing of this PR with conjunction of graceful shutdown and the noticed the following behavior:

During shutdown, I noticed that the first status update would be sent which updated the [message and reason](https://github.com/kubernetes/kubernetes/blob/9e5efeb/pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go#L323-L324) for the pod. The message was set to `Terminated`, which makes it look like the pod is already terminated, but the phase is still running. Only after all the containers terminate (which is terminationGracePeriod seconds later), will the second status update be sent with `Phase=Failed`. 

This is a bit confusing, since the pod can be stuck with `Terminated` message, but still be running if the node happens to shutdown before the second status update is sent. This is also behavior change, since previously phase=Failed and message/reason updates were sent in a single update request.

To fix it, perhaps we should also apply the same logic of using the oldPhase to the message and reason.

smarterclayton replied:
Interesting.  So this is two factors competing:

1. As a user, I want to know that kubelet is responding (post this change, it takes longer to know that)
2. As a user, I want to be able to see consistent messaging (post this change, message/reason and phase were not aligned)

The best option (in a follow up) is to have a couple of new conditions - `Terminating` and `Evicted` that capture one way transitions (once the condition is present, there is only one outcome) and cleanly solve 1.  Since that's a longer change, I agree that message and reason should probably be aligned in this pr.  I could foresee some user regression in that (especially for eviction) but that's just a reason to prioritize the conditions.

bobbypage replied:
Makes sense. I agree long term we probably need an intermediate state (i.e. `Terminating...`), but for now I think it'll best to align message and reason in a single update.

rphillips commented:
Updating the PR, the OpenShift CI passed after a pointer from Clayton.

rphillips commented:
Is the PR ready for review?

smarterclayton commented:
Yes for review, although I want to follow the two things David mentioned.

smarterclayton commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-e2e-gce-ubuntu-containerd/1501654012359872512/

```
Kubernetes e2e suite: [sig-network] KubeProxy should set TCP CLOSE_WAIT timeout [Privileged] expand_less | 49s
-- | --
test/e2e/network/kube_proxy.go:52 Mar  9 20:53:47.022: Unexpected error:     <*errors.errorString \| 0xc0001113c0>: {         s: "pod ran to completion",     }     pod ran to completion occurred test/e2e/framework/pods.go:107
```

is "out of pods", which might be due to parallelism.  Investigating

bobbypage commented on test/e2e/apps/job.go:
```diff
@@ -361,6 +367,52 @@ var _ = SIGDescribe("Job", func() {
 		}
 	})
+
+	ginkgo.It("should run a job to completion with cpu requests", func() {
```
Perhaps we need to add a [Serial] here to address https://github.com/kubernetes/kubernetes/pull/108366#issuecomment-1064547192 so this test don't run in parallel

smarterclayton replied:
Good point

smarterclayton committed:
test: Add E2E for init container pod deletion

Exploring termination revealed we have race conditions in certain
parts of pod initialization and termination. To better catch these
issues refactor the existing test so it can be reused, and then test
a number of alternate scenarios.

smarterclayton commented:
/test pull-kubernetes-node-kubelet-serial-containerd

bobbypage commented:
Thanks for the updates! Looks good now (all tests are passing) and the new added test from https://github.com/kubernetes/kubernetes/pull/108599 is passing in `pull-kubernetes-node-kubelet-serial-containerd` while it [isn't](https://github.com/kubernetes/kubernetes/pull/108599#issuecomment-1063378852) in master, demonstrating this change is also fixing https://github.com/kubernetes/kubernetes/issues/108594. 

Can we add the Serial tag to the test (i.e. https://github.com/kubernetes/kubernetes/pull/108366#discussion_r824226896) so it will not impact other presubmit tests? Anything else needed? It looks very close to being ready, thanks for all your work on this @smarterclayton.

smarterclayton commented:
Updated

/test pull-kubernetes-node-kubelet-serial-containerd

smarterclayton commented:
/test all

bobbypage commented:
`pull-kubernetes-node-kubelet-serial-containerd` [failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-node-kubelet-serial-containerd/1504161025379799040/
) looks to be unrelated flake

```
E2eNode Suite: [sig-node] Density [Serial] [Slow] create a batch of pods latency/resource should be within limit when create 10 pods with 0s interval expand_less | 1m43s
-- | --
{ Failure test/e2e_node/density_test.go:104 Mar 16 18:44:13.995: CPU usage exceeding limits:  node n1-standard-2-ubuntu-gke-2004-1-20-v20220316-e95e6759:  container "runtime": expected 95th% usage < 0.600; got 0.618 test/e2e_node/resource_usage_test.go:214}
```

bobbypage commented:
/test pull-kubernetes-node-kubelet-serial-containerd

liggitt commented:
> `pull-kubernetes-node-kubelet-serial-containerd` [failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108366/pull-kubernetes-node-kubelet-serial-containerd/1504161025379799040/) looks to be unrelated flake

if that repeats, could also see if the same failure exists in post-submits or on a no-op PR against master

liggitt commented:
/test pull-kubernetes-node-kubelet-serial-containerd

liggitt commented:
serial jobs/tests to watch post-merge:
* e2e: https://testgrid.k8s.io/google-gce#gce-cos-master-serial&width=20&sort-by-failures=&include-filter-by-regex=apps (though the added `[sig-apps] Job should run a job to completion with cpu requests` test was passing in the presubmit e2e before the `[Serial]` tag was added, so I'm not that concerned about that one)
* node-e2e: https://testgrid.k8s.io/sig-node-containerd#cos-cgroupv2-containerd-node-e2e-serial&width=20&include-filter-by-regex=GracefulNodeShutdown

liggitt commented:
green run of `pull-kubernetes-node-kubelet-serial-containerd` 🎉

bobbypage commented:
/lgtm

Thanks Clayton again for all your work on this, really appreciate it!

/hold

Adding hold in case any additional approval needed from your side, please un-hold when you're ready.

rphillips commented:
I'm running the OpenShift CI against the changes located [here](https://github.com/openshift/kubernetes/pull/1199#issuecomment-1069768544).

Let's plan on merging this on Thursday if the CI run passes.

liggitt commented:
> I'm running the OpenShift CI against this PR [here](https://github.com/openshift/kubernetes/pull/1199#issuecomment-1069768544).
> 
> Let's plan on merging this on Thursday if the CI run passes.

sgtm. what version of Kubernetes is that testing against? @bobbypage has picks of this open against release-1.22 (https://github.com/kubernetes/kubernetes/pull/108749) and release-1.23 (https://github.com/kubernetes/kubernetes/pull/108723) that are passing K8s presubmit CI, and he has validated against serial e2es manually. It would be useful to validate those against the openshift versions based on those releases as well if you have the ability to trigger that.

liggitt commented:
> I'm running the OpenShift CI against this PR [here](https://github.com/openshift/kubernetes/pull/1199#issuecomment-1069768544).

@rphillips - looks like `ci/prow/k8s-e2e-gcp` and `ci/prow/k8s-e2e-gcp-serial` succeeded on that PR, but a lot of the openshift jobs referenced at https://github.com/openshift/kubernetes/pull/1199#issuecomment-1069768544 failed with issues unrelated to the change ("Start a service that hosts the RPMs generated by this build")

Given the healthy presubmits here (including the new e2e and node-e2e tests added specifically exercising the regression), I'd be comfortable merging this and watching the [k8s periodic jobs](https://github.com/kubernetes/kubernetes/pull/108366#issuecomment-1069721995) for healthy serial runs before merging the release branch picks

rphillips commented:
Those jobs are running against 1.23. 

+1 we can merge with those successes.

/hold cancel

rphillips commented:
/triage accepted
/priority critical-urgent

liggitt commented:
weird, `pull-kubernetes-e2e-gce-100-performance` passed 10 minutes ago but isn't reflected here

rphillips commented:
@liggitt GitHub is having an outage right now...

liggitt commented:
green runs of both serial e2e and serial node e2e post-merge 🎉

liggitt commented:
multiple green runs of the serial jobs now (the e2e serial failures were either a prow pod timeout error unrelated to the job, or an [pre-existing flake](https://storage.googleapis.com/k8s-triage/index.html?text=ssh%3A%20handshake%20failed&job=serial) sshing to the node)

I'm comfortable with unholding the release branch picks at this point
